Components/FlightControl/RocketSM.cpp

```diff
@@ -0,0 +1,455 @@
+/**
+ ******************************************************************************
+ * File Name          : RocketSM.cpp
+ * Description        : Primary rocket state machine
+ ******************************************************************************
+*/
+#include "RocketSM.hpp"
+#include "SystemDefines.hpp"
+/* Rocket Control ------------------------------------------------------------------*/
+/**
+ * @brief Default constructor for Rocket SM, initializes all states
+ */
+RocketControl::RocketControl()
+{
+    // Setup the internal array of states. Setup in order of enum.
+    stateArray[RS_PRELAUNCH] = new PreLaunch();
+    stateArray[RS_FILL] = new Fill();
+    stateArray[RS_ARM] = new Arm();
+    stateArray[RS_IGNITION] = new Ignition();
+    stateArray[RS_LAUNCH] = new Launch();
+    stateArray[RS_BURN] = new Burn();
+    stateArray[RS_COAST] = new Coast();
+    stateArray[RS_DESCENT] = new Descent();
+    stateArray[RS_RECOVERY] = new Recovery();
+    stateArray[RS_ABORT] = new Abort();
+
+    // Verify all states are initialized
+    for (uint8_t i = 0; i < RS_NONE-1; i++)
+        SOAR_ASSERT(stateArray[i] != nullptr);
+}
+
+/**
+ * @brief Handles state transitions
+ * @param nextState The next state to transition to
+ * @return The state after the transition
+ */
+RocketState RocketControl::TransitionState(RocketState nextState)
+{
+    // Check if we're already in the next state (TransitionState does not allow entry into the existing state)
+    if (nextState == rs_currentState->GetStateID())
+        return rs_currentState->GetStateID();
+
+    // Check the next state is valid
+    if (nextState >= RS_NONE)
+        return rs_currentState->GetStateID();
+
+    // Exit the current state
+    rs_currentState->OnExit();
+
+    // Set the next state
+    rs_currentState = stateArray[nextState];
+
+    // Assert the next state is initalized
+    SOAR_ASSERT(rs_currentState != nullptr, "rs_currentState is nullptr in TransitionState");
+
+    // Enter the current state
+    rs_currentState->OnEnter();
+
+    // Return the state after the transition
+    return rs_currentState->GetStateID();
+}
+
+/**
+ * @brief Handles current command
+ * @param cm The command to handle
+ */
+void RocketControl::HandleCommand(Command& cm)
+{
+    SOAR_ASSERT(rs_currentState != nullptr, "Command received before state machine initialized");
+
+    // Handle the command based on the current state
+    rs_currentState->HandleCommand(cm);
+}
+
+/* Base State ------------------------------------------------------------------*/
+/**
+ * @brief General handler for actions that should be supported by all rocket state machines
+ */
+RocketState BaseRocketState::HandleGeneralStateCommands(RocketControlCommands rcAction)
```

Comment:
Im a bit confused with when these Handle_X_StateCommands() will be used. Let me know if this is correct: each state has a state-specific HandleCommand() function and this is where we block transitions to invalid/unreachable states. (how is this chained with RocketControl::TransitionState()?). Only prelaunch can call HandleNonIgnitionCommands (which, by the way, is probably a general task that should be implemented by the base class along with sanity checks using rsStateID). We also only call HandleGeneralStateCommands() in prelaunch, but it's called at the start along with the other command checks...?

Reply:
Any given HandleCommand function **only** implements functionality that specific state is allowed to process. If a specific set of functionality can be shared between multiple states we implement that by deriving that state and calling the parent class' unique function (to avoid code duplication). HandleNonIgnitionCommands is, as it sounds, "a set of commands that can be done by any state prior to ignition eg. Abort"

This specific function was implemented based on my previous assumption that we'd have any sort of security on commands from being executed in a given state... but given that we're apparently going to allow everything from every state anyway..... this can likely be folded into a general "HandleGeneralCommands" function, with only abort being handle d in this particular function.